components/core/tests/test-ffi_KeyValuePairLogEvent.cpp

```diff
@@ -0,0 +1,465 @@
+
+        for (auto const& [node_id, optional_value] : invalid_node_id_value_pairs) {
+            KeyValuePairLogEvent::NodeIdValuePairs node_id_value_pair_to_test;
+            if (optional_value.has_value()) {
+                node_id_value_pair_to_test.emplace(node_id, optional_value.value());
+            } else {
+                node_id_value_pair_to_test.emplace(node_id, std::nullopt);
+            }
+            auto const result{KeyValuePairLogEvent::create(
+                    schema_tree,
+                    std::move(node_id_value_pair_to_test),
+                    UtcOffset{0}
+            )};
+            REQUIRE(result.has_error());
+            auto const& err{result.error()};
+            REQUIRE((std::errc::protocol_error == err));
+        }
+    }
+
+    SECTION("Test valid id-value pairs") {
+        KeyValuePairLogEvent::NodeIdValuePairs valid_node_id_value_pairs;
+        /*
+         * The sub schema tree of `valid_node_id_value_pairs`:
+         * <0:root:Obj>
+         *      |
+         *      |------------> <1:a:Obj>
+         *      |                  |
+         *      |--> <2:a:Int>     |--> <3:b:Obj>
+         *                                  |
+         *                                  |------------> <4:c:Obj>
+         *                                  |                  |
+         *                                  |--> <5:d:Str>     |--> <7:a:UnstructuredArray>
+         *                                  |                  |
+         *                                  |                  |--> <8:d:Str>
+         *                                  |                  |
+         *                                  |--> <10:e:Obj>    |
+         *                                                     |
+         *                                                     |--> <11:f:Obj>
+         */
+        // NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+        valid_node_id_value_pairs.emplace(2, Value{static_cast<value_int_t>(0)});
+        valid_node_id_value_pairs.emplace(5, Value{string{"Test"}});
+        valid_node_id_value_pairs.emplace(
+                8,
+                Value{get_encoded_text_ast<four_byte_encoded_variable_t>(cStringToEncode)}
+        );
+        valid_node_id_value_pairs.emplace(
+                7,
+                Value{get_encoded_text_ast<eight_byte_encoded_variable_t>(cStringToEncode)}
+        );
+        valid_node_id_value_pairs.emplace(10, Value{});
+        valid_node_id_value_pairs.emplace(11, std::nullopt);
+        // NOLINTEND(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+        auto const result{
+                KeyValuePairLogEvent::create(schema_tree, valid_node_id_value_pairs, UtcOffset{0})
+        };
+        REQUIRE_FALSE(result.has_error());
+
+        SECTION("Test duplicated key conflict on node #3") {
+            auto conflict_pairs{valid_node_id_value_pairs};
```

Comment:
I thought the idea of using nested sections (besides breaking the code up) is that you wouldn't need to duplicate `valid_node_id_value_pairs` since it would start with a fresh state in each section, no? (You'd obviously call it something other than `valid_node_id_value_pairs`, but still.